File: server/src/pkm/server/daily.py
# pattern: Functional Core
"""The server-side pieces of the daily/journal machinery: the date
window, the emptiness test for the empty-daily cleanup, and which days a
journal batch shows. The date <-> title spelling itself lives in
`pkm.contracts.daily`, shared with the CLI and MCP server."""
from __future__ import annotations

from collections.abc import Iterable, Sequence
from datetime import date, timedelta


def past_week_dates(today: date) -> list[date]:
    """The 7 dates before `today`, newest first. `today` itself is excluded:
    the journal auto-creates today's page for composing."""
    return [today - timedelta(days=i) for i in range(1, 8)]


def is_page_empty(texts: Sequence[str]) -> bool:
    """True when every block text is empty/whitespace (or there are none)."""
    return all(not t.strip() for t in texts)


def select_journal_days(nonempty: Iterable[date], today: date,
                        before: date | None, limit: int) -> list[date]:
    """The dates a journal batch shows, newest first (pkm-03x6). With no
    cursor: today leads — always, even when empty, so there is a page to
    compose into — followed by the most recent non-empty days before it.
    With a `before` cursor: the most recent non-empty days strictly before
    it. Empty gap days are omitted entirely; a batch shorter than `limit`
    tells the client the journal is exhausted."""
    if before is None:
        pool = sorted((d for d in nonempty if d < today), reverse=True)
        return [today, *pool[:max(0, limit - 1)]]
    pool = sorted((d for d in nonempty if d < before), reverse=True)
    return pool[:limit]

File: web/src/replica/daily.ts
// pattern: Functional Core
// TS port of server daily.py title helpers: Roam's ordinal daily-page
// titles ("July 13th, 2026"). Local daily auto-create offline needs the
// exact same format the server generates.

const MONTHS = ["January", "February", "March", "April", "May", "June",
                "July", "August", "September", "October", "November",
                "December"];

const TITLE_RE = new RegExp(
  `^(${MONTHS.join("|")}) (\\d{1,2})(st|nd|rd|th), (\\d{4})$`);

function suffix(day: number): string {
  if (day % 100 >= 10 && day % 100 <= 20) return "th";
  return { 1: "st", 2: "nd", 3: "rd" }[day % 10] ?? "th";
}

export function titleForDate(d: Date): string {
  return `${MONTHS[d.getMonth()]} ${d.getDate()}${suffix(d.getDate())},` +
    ` ${d.getFullYear()}`;
}

export function dateForTitle(title: string): Date | null {
  const m = TITLE_RE.exec(title);
  if (!m) return null;
  const day = Number(m[2]);
  if (suffix(day) !== m[3]) return null;
  return new Date(Number(m[4]), MONTHS.indexOf(m[1]), day);
}

// TS port of daily.py select_journal_days (pkm-03x6): the dates a journal
// batch shows, newest first. No cursor: today leads — always, even when
// empty, so there is a page to compose into — then the most recent
// non-empty days before it. With a cursor: the most recent non-empty days
// strictly before it. All dates are local midnights.
export function selectJournalDays(nonempty: Date[], today: Date,
                                  before: Date | null, limit: number): Date[] {
  const pool = nonempty
    .filter((d) => d.getTime() < (before ?? today).getTime())
    .sort((a, b) => b.getTime() - a.getTime());
  if (before === null) {
    return [today, ...pool.slice(0, Math.max(0, limit - 1))];
  }
  return pool.slice(0, limit);
}
